crates/iroha_core/src/sumeragi/main_loop.rs

```diff
@@ -841,57 +839,74 @@ impl Sumeragi {
 
-            let mut state_block = state.block();
-            let create_block_start_time = Instant::now();
-            let new_block = BlockBuilder::new(transactions)
-                .chain(self.topology.view_change_index(), &mut state_block)
-                .sign(self.key_pair.private_key())
-                .unpack(|e| self.send_event(e));
-
-            let created_in = create_block_start_time.elapsed();
-            let pipeline_time = state.world.view().parameters().sumeragi.pipeline_time();
-            if created_in > pipeline_time / 2 {
-                warn!(
-                    role=%self.role(),
-                    peer_id=%self.peer_id,
-                    "Creating block takes too much time. \
-                    This might prevent consensus from operating. \
-                    Consider increasing `commit_time` or decreasing `max_transactions_in_block`"
-                );
+            let pre_signed_block = BlockPayload {
+                header: BlockBuilder::make_header(
+                    state.view().latest_block().as_deref(),
+                    self.topology.view_change_index(),
+                    &transactions,
+                    state
+                        .view()
+                        .world
+                        .parameters()
+                        .sumeragi
+                        .consensus_estimation(),
+                ),
+                transactions,
             }
+            .sign(self.key_pair.private_key());
 
+            let block_created_msg = BlockCreated {
+                block: pre_signed_block,
+            };
             if self.topology.is_consensus_required().is_some() {
-                info!(
-                    peer_id=%self.peer_id,
-                    block=%new_block.as_ref().hash(),
-                    view_change_index=%self.topology.view_change_index(),
-                    txns=%new_block.as_ref().transactions().len(),
-                    created_in_ms=%created_in.as_millis(),
-                    "Block created"
-                );
+                self.broadcast_packet(block_created_msg.clone());
+            }
 
-                let msg = BlockCreated::from(&new_block);
-                *voting_block = Some(VotingBlock::new(new_block, state_block));
-                self.broadcast_packet(msg);
+            info!(
+                peer_id=%self.peer_id,
+                view_change_index=%self.topology.view_change_index(),
+                block_hash=%block_created_msg.block.hash(),
+                txns=%block_created_msg.block.transactions().len(),
+                "Block created"
+            );
+
+            let new_voting_block = self
+                .validate_block(
+                    state,
+                    &self.topology,
+                    genesis_account,
+                    block_created_msg,
+                    voting_block,
+                )
+                .expect("We just created this block ourselves, it has to be valid.");
```

Comment:
I think this place might panic if any transaction is invalid, isn't it?

Also validators would reject block if any transaction is invalid.

Reply:
My understanding is that the transaction should have been rejected by the queue already. If it were true that a transaction could fail the block. Then we could be perpetually failing to generate a block in all rounds because the bad transaction would be gossiped around to all peers. This would be true already. Therefore it is much safer for our sanity to crash here because we should never be constructing a bad block ourselves.

Reply:
You can check `validate_keep_voting_block` and `validate_transactions` inside it.
If there is mismatch in transaction error (and we set it to `None` despite it's actual status) we got rejection.
It's not possible to fully validate transaction in the queue without executing it.

Reply:
We want to categorize transactions and reject a block if it's `CommittedTransaction::error` is incorrect from what the current peer's evaluation produced

Reply:
@erigara but how can there be a mismatch in the transaction error when the received block has it set to None for all transactions?

Reply:
Look [here](https://github.com/hyperledger/iroha/blob/002803facb9d2ae9fb414f5428e6356510fe094c/crates/iroha_core/src/block.rs#L619), if transaction has `None` but after execution we discovered that it not actually `None` error is returned

Reply:
> Look [here](https://github.com/hyperledger/iroha/blob/002803facb9d2ae9fb414f5428e6356510fe094c/crates/iroha_core/src/block.rs#L619), if transaction has `None` but after execution we discovered that it not actually `None` error is returned   

Shouldn't we remove that check there?

Reply:
I have removed that check and replaced it with a mutation of the error value in the block.